<a href="https://colab.research.google.com/github/JonNData/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/Jon_LS_DS12_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [1]:
### YOUR CODE STARTS HERE
column_headers = ['party', 'handicapped-infants', 'water-project', 
'budget', 'physician-fee-freeze', 'el-salvador-aid', 'religious-groups', 
'anti-satellite-ban', 'aid-to-contras', 'mx-missile', 'immigration', 
'synfuels', 'education', 'right-to-sue', 'crime', 'duty-free', 
'south-africa']

import pandas as pd
import numpy as np

df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data', names= column_headers)
print(df.shape)
df.head()

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [3]:
df = df.replace({ '?' : np.NaN, 'n': 0, 'y' : 1})
df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [4]:
df.isnull().sum()

party                     0
handicapped-infants      12
water-project            48
budget                   11
physician-fee-freeze     11
el-salvador-aid          15
religious-groups         11
anti-satellite-ban       14
aid-to-contras           15
mx-missile               22
immigration               7
synfuels                 21
education                31
right-to-sue             25
crime                    17
duty-free                28
south-africa            104
dtype: int64

In [5]:
rep = df[df.party == 'republican']
rep.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

In [6]:
rep.describe()

handicapped-infants  water-project  ...   duty-free  south-africa
count           165.000000     148.000000  ...  156.000000    146.000000
mean              0.187879       0.506757  ...    0.089744      0.657534
std               0.391804       0.501652  ...    0.286735      0.476168
min               0.000000       0.000000  ...    0.000000      0.000000
25%               0.000000       0.000000  ...    0.000000      0.000000
50%               0.000000       1.000000  ...    0.000000      1.000000
75%               0.000000       1.000000  ...    0.000000      1.000000
max               1.000000       1.000000  ...    1.000000      1.000000

[8 rows x 16 columns]

In [7]:
dem = df[df['party'] == 'democrat' ]
dem.describe()

handicapped-infants  water-project  ...   duty-free  south-africa
count           258.000000     239.000000  ...  251.000000    185.000000
mean              0.604651       0.502092  ...    0.637450      0.935135
std               0.489876       0.501045  ...    0.481697      0.246956
min               0.000000       0.000000  ...    0.000000      0.000000
25%               0.000000       0.000000  ...    0.000000      1.000000
50%               1.000000       1.000000  ...    1.000000      1.000000
75%               1.000000       1.000000  ...    1.000000      1.000000
max               1.000000       1.000000  ...    1.000000      1.000000

[8 rows x 16 columns]

In [0]:
from scipy.stats import ttest_ind, ttest_ind_from_stats, ttest_rel

In [12]:
# 2. Find dems more than rep issue
# Ho : The mean of dems.handicapped-infants <= the mean of rep.handicapped-infants
# Ha : The mean of dems.handicapped-infants > the mean of rep.handicapped-infants
ttest_ind(rep['handicapped-infants'], dem['handicapped-infants'], nan_policy='omit')

Ttest_indResult(statistic=-9.205264294809222, pvalue=1.613440327937243e-18)

In [0]:
# Since T = -9.205 and p-value/2 = (1.6e-18)/2 = 8e-19, we are 95% confident we
# can reject the null hypothesis that  dems.handicapped-infants are less than or
# equal to rep.handicapped-infants


## Since T = -9.205 and p-value/2 = (1.6e-18)/2 = 8e-19, we are 95% confident we can reject the null hypothesis that  dems.handicapped-infants are less than or equal to rep.handicapped-infants. Democrats are more likely to vote yes for the handicapped-infants issue than republicans are.

In [15]:
# 3. Find rep more than dems issue
# Ho : the mean of rep['el-salvador-aid	'] <= the mean of dem['el-salvador-aid']
# Ha : the mean of rep['el-salvador-aid	'] > the mean of dem['el-salvador-aid']
ttest_ind(dem['el-salvador-aid'], rep['el-salvador-aid'], nan_policy='omit')

Ttest_indResult(statistic=-21.13669261173219, pvalue=5.600520111729011e-68)

In [0]:
# I did the parameters in the reverse order to challenge my understanding of 
# a negative t-statistic. 

## Since T = -21.14 and p-value/2 = (5.6e-68/2) = 2.8e-68, we are 95% confident that we can reject the null hypothesis that republicans are less than or equal in favor of el-salvador-aide. Republicans are more likely to vote yes on the aid to El Salvador than democrats

In [16]:
# 4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 
# Ho : the mean of rep['immigration'] =  the mean of dem['immigration']
# Ha : the mean of rep['immigration'] !=  the mean of dem['immigration']

ttest_ind(rep['immigration'], dem['immigration'], nan_policy='omit')

Ttest_indResult(statistic=1.7359117329695164, pvalue=0.08330248490425066)

In [17]:
# No such luck, let's try water-project
# Ho : the mean of rep['water-project'] =  the mean of dem['water-project']
# Ha : the mean of rep['water-project'] !=  the mean of dem['water-project']

ttest_ind(rep['water-project'], dem['water-project'], nan_policy='omit')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

## Since the T-statistic = .089 and the p-value= 0.929 we fail to reject the null hypothesis that the mean of rep['water-project'] =  the mean of dem['water-project']. There is no statistically significant difference between parties voting on the water-project issue.